follow_fork/follow_fork.c

```diff
@@ -0,0 +1,123 @@
+#define _POSIX_SOURCE
+#define _GNU_SOURCE
+
+#include "undolr.h"
+#include "undolr_deprecated.h"
+
+#include <errno.h>
+#include <stdio.h>
+#include <sys/types.h>
+#include <unistd.h>
+#include <string.h>
+#include <stdbool.h>
+#include <time.h>
+#include <stdlib.h>
+#include <dlfcn.h>
+#include <linux/limits.h>
+
+extern char *__progname;
+
+void attach_lr (int, char**, char**) __attribute__ ((constructor));
+
+static bool
+is_valid_process(pid_t pid)
+{
+    if (strncmp(__progname, "undo", 4) == 0)
+    {
+        if (unsetenv("LD_PRELOAD") == -1)
+        {
+            perror("unsetenv");
+            exit(1);
+        }
+        return false;
+    }
+    return true; 
+}
+
+static void*
+get_sym_addr(const char *sym)
+{
+    void *fptr = NULL;
+    
+    fptr = dlsym(RTLD_NEXT, sym);
+    if (fptr == NULL)
+    {
+        fprintf(stderr, "Error: %s, calling dlsym with %s\n",
+                        dlerror(), sym);
+        abort();
+    }
+    return fptr;
+}
+
+static bool
+start_recording(pid_t pid)
+{
+    undolr_error_t err;
+    int e = undolr_start(&err);
+    if (e)
+    {
+        fprintf(stderr, "undolr_recording_start failed: err=%i, errno=%i\n", err, errno);
+        return false;
+    }
+
+    time_t cur_time = time(0);
+    char rec_fname[PATH_MAX] = {0};
+
+    // Recording files are saved in /tmp, change this if you want them saved somewhere else
+    snprintf(rec_fname, PATH_MAX, "/tmp/%s_%d_%lu.undo", __progname, pid, cur_time);
+
+    e = undolr_save_on_termination(rec_fname);
+    if (e)
+    {
+        fprintf(stderr, "undolr_save_on_termination failed: errno=%i\n", errno);
+        return false;
+    }
+
+    return true;
+}
+
+
+/* Interpose fork to follow forked processes */
+pid_t
+fork(void)
+{
+    typedef pid_t (*fork_function)();
+    static fork_function real_fork = NULL;
+
+    if (!real_fork)
+    {
+        real_fork = (fork_function)get_sym_addr("fork");
+    }
+
+    pid_t pidx = real_fork();
+    if (pidx == 0) 
+    {
+        pid_t cpid = getpid();
+        if (is_valid_process(cpid))
+        {
+            // Start recording the child
+            if (!start_recording(cpid))
+            {
+                fprintf(stderr, "start_recording() failed\n");
+            }
+        }
+    }
+    return pidx;
+}
+
+void
+attach_lr(int argc, char **argv, char **envp)
```

Comment:
How does this work?
Why are there two ways to attach LR?

Reply:
Good question, I'll need to dig a bit deeper to understand Nikilesh's intentions.

Reply:
This is a `__attribute__ ((constructor))` so it records the first process, then the `fork` records the others.
I think this would be clearer if the declaration to add the attribute were just before this function and a comment explained it.

Reply:
Added description and moved the declaration closer to the definition.